### PR TITLE
Allow (B|R)VALID at the same time as (AW|W|AR)READY

### PR DESCRIPTION
**Pull request by leastrobino**
- Write transaction: change the order of RisingEdge() and ReadOnly() to read BVALID in the same clock cycle as the last of AWREADY and WREADY
- Read transaction: change the order of RisingEdge() and ReadOnly() to read RVALID in the same clock cycle as ARREADY
- Use lock on read transaction

### CODE REVIEW

**Comment by imphil** [comment]
Thanks for your patch! 

I'm not an AXI expert, but nobody voiced objection and the commit looks fine, so there's no reason to keep this open any longer.

File: cocotb/drivers/amba.py [hunk 1]
```diff
@@ -146,11 +144,11 @@ def write(self, address, value, byte_enable=0xf, address_latency=0,
 
         # Wait for the response
         while True:
-            yield ReadOnly()
             if self.bus.BVALID.value and self.bus.BREADY.value:
                 result = self.bus.BRESP.value
                 break
             yield RisingEdge(self.clock)
+            yield ReadOnly()
```
Comment (eric-wieser) [inline cocotb/drivers/amba.py]: This change doesn't look right to me - we shouldn't be looking at signals until the `ReadOnly` phase

**Comment by eric-wieser** [comment]
@leastrobino, can you elaborate on the reasoning behind the first two bullet points? It's my understanding that yielding `ReadOnly` while already in a readonly phase does not advance time, so I'm not sure why adding or removing it would affect which clock cycle a read happens in.

What simulator were you using when you had this trouble? It's also possible that the behavior of these triggers varies by simulator.

**Comment by leastrobino** [comment]
@eric-wieser, I proposed those changes because some IPs I have in my design behave like that, BVALID or RVALID is set at the same time as the other channels READY signals.

Although accepted by ARM processors and/or Xilinx AXI Interconnect, it seems this behavior is slightly off-standard according to AMBA AXI and ACE Protocol Specification, A3.3.1:
- Read transaction dependencies:
  The slave must wait for both ARVALID and ARREADY to be asserted before it asserts RVALID to indicate that valid data is available.
- AXI4 and AXI5 write transaction dependencies:
  The slave must wait for AWVALID, AWREADY, WVALID, and WREADY to be asserted before asserting BVALID.

After having read more thoroughly the standard, I propose to revert these changes and keep only the changes on read transaction lock (pull request #1283).

**Comment by eric-wieser** [comment]
Thanks for the analysis - I think I understand your intent better now, but I don't think this patch correctly implements it. We'll get this reverted before the 1.3 release.

**Pull request merged**